### PR TITLE
Directly use Best Possible State to calculate DifferenceWithIdealStateGauge metrics instead of relying on the persisted IdealState.

### LINKED ISSUE #1670: IdealState does not appear to be written
### Describe the bug
There appears to be a bug with either the calculation of the "DifferenceWithIdealStateGauge" metric, or the writing of ideal state.  We recently went through an upgrade from Helix 0.6.9 to 1.0.1.  When viewing the current state of Helix `--listResourceInfo`, IdealState in both 0.6.9 and 1.0.1 appears to always be empty.  However, the metric calculation seems to have changed between the two versions.  In 0.6.9, "DifferenceWithIdealStateGauge" would be emitted as 0.  In 1.0.1, it will correctly emit an accurate metric as an empty IdealState with all partitions accurately assigned will always emit the total partitions as the "DifferenceWithIdealStateGauge" metric.

// Metric calculation in 1.0.1
[(1.0.1) ResourceMonitor.java](https://github.com/apache/helix/blob/helix-1.0.1/helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ResourceMonitor.java#L243)

// Metric calculation in 0.6.9
[(0.6.9) ResourceMonitor.java](https://github.com/apache/helix/blob/helix-0.6.9/helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ResourceMonitor.java#L98)

This could potentially be related to https://github.com/apache/helix/issues/609.

### To Reproduce
https://github.com/jeff-blaisdell/helix-issue contains a full working example.  Greater details on environment / cluster setup can be found in project README.

### Expected behavior
DifferenceWithIdealStateGauge should correctly report 0.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

Resolve #1670 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This PR aims to resolve the undesired dependencies between DifferenceWithIdealStateGauge metric and the PERSIST_XXXXXX_ASSIGNMENT configuration. Before the change, if the assignment is not persisted, then the metric will report incorrect data.
To fix the issue, the calculation must be moved from the ExternalViewComputeStage to the BestPossibleStateCalcStage. Because the Best Possible State is only available in the BestPossibleStateCalcStage assuming the PERSIST_XXXXXX_ASSIGNMENT option is not turned on. In addition to the main logic migration, other changes listed following are required to ensure multi-thread safety.
1. Concurrent control of the EV cache since it will now be read in the BestPossibleStateCalcStage in addition to the ExternalViewComputeStage.
2. Minor changes in the diff computing logic for the corner cases since the Best Possible State state mapping is not exactly the same as IdealState persist assignment.
3. Cleanup PersistAssignmentStage logic so it won't modify the IdealState cache anymore. It is supposed to be read-only.
4. Test cases are modified to cover the changes. Also, the metric is now reporting correct results in some corner cases such as all the nodes are disabled. The test cases which are testing based on the wrong behavior have been fixed.

### Tests

- [X] The following tests are written for this issue:

Modify TestClusterAggregateMetrics.java according to the new behavior.
Adding new checks in TestClusterStatusMonitor.java to validate the DifferenceWithIdealStateGauge result.

- The following is the result of the "mvn test" command on the appropriate module:

[INFO] Tests run: 1264, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5,080.412 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1264, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:24 h
[INFO] Finished at: 2021-04-13T20:15:05-07:00
[INFO] ------------------------------------------------------------------------

Note that TestCleanupExternalView is not stable currently. It frequently fails in the master branch too. We will take care of that separately.

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java [hunk 1]
```diff
@@ -128,25 +130,30 @@ private void persistAssignment(final Resource resource, final ResourceController
...
+            ZNRecord deltaRecord = delta.getRecord();
+            // Overwrite MapFields and ListFields items with the same key.
+            if (!deltaRecord.getMapFields().isEmpty()) {
+              // Note that default merge will keep old values in the maps unchanged, which is not desired.
               current.getMapFields().clear();
-              current.getMapFields().putAll(idealState.getRecord().getMapFields());
-              current.getListFields().putAll(idealState.getRecord().getListFields());
+              current.getMapFields().putAll(deltaRecord.getMapFields());
+            }
+            if (!deltaRecord.getListFields().isEmpty()) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java:149]: Question: why we need this check? We can add lists field anyway just not clear it up, right?

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java [hunk 2]
```diff
@@ -205,7 +205,9 @@ private void refreshExternalViews(final HelixDataAccessor accessor) {
     // As we are not listening on external view change, external view will be
     // refreshed once during the cache's first refresh() call, or when full refresh is required
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.EXTERNAL_VIEW).getAndSet(false)) {
-      _externalViewCache.refresh(accessor);
+      synchronized (_externalViewCache) {
+        _externalViewCache.refresh(accessor);
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java:209]: That would be too heavy. And I'm almost 100% sure it will impact performance since the cache is used everywhere in the pipeline.

File: helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java (hunk 1 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java:149]: This is for backward compatibility. Not adding it will be fine, logic-wise. But it will change the behavior, so some tests fail. This is because the change in this PR changes the input.

Since adding this check also makes sense to me, I decide to keep the same behavior.
What's your opinion?

Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java:149]: I agree better to keep the behavior. But the original code does not perform the check, right? The check changed the behavior of our code that only list field contains something then we add. 

Either way is fine. I dont have strong preference because the only thing I can think about is custom mode / non rebalanced resource does not have the list field.  For non rebalanced resource, it will be eventually caught up if the it get rebalanced with list field fulfilled.

**Review by dasahcc — APPROVED** [review]
overall, lgtm

File: helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java (hunk 1 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java:149]: Since I changed the logic, if I pass an empty map through the "deltaRecord", it means don't update. Originally it means the IdealState should have the empty map. So the meaning of the input has been changed. So this check will ensure the behavior keeps the same.

**Comment by jiajunwang** [comment]
Approved by @dasahcc, I will merge later.
